ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/ExecutionKeyFactory.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright 2019 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License; you may not use this file
+ * except in compliance with the License.  You may obtain a copy of the License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.plan;
+
+import io.confluent.ksql.execution.builder.KsqlQueryBuilder;
+import io.confluent.ksql.execution.context.QueryContext;
+import io.confluent.ksql.schema.ksql.PhysicalSchema;
+import io.confluent.ksql.serde.FormatInfo;
+import io.confluent.ksql.serde.WindowInfo;
+import io.confluent.ksql.testing.EffectivelyImmutable;
+import java.util.Objects;
+import org.apache.kafka.common.serialization.Serde;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.streams.kstream.Windowed;
+
+/**
+ * The {@code ExecutionKeyFactory} is in charge of creating the keys
+ * and the Serdes for the keys of any particular execution step.
+ *
+ * @param <K> the type of the key, usually either {@code Struct}
+ *            or {@code Windowed<Struct>}
+ */
+@EffectivelyImmutable
+public interface ExecutionKeyFactory<K> {
+
+  Serde<K> buildKeySerde(
+      FormatInfo format,
+      PhysicalSchema physicalSchema,
+      QueryContext queryContext
+  );
+
+  /**
+   * @return a new {@code ExecutionKeyFactory}
+   */
+  ExecutionKeyFactory<K> withQueryBuilder(KsqlQueryBuilder builder);
+
+  /**
+   * This method can construct a new key given the contents of the old key and the
+   * desired Struct representation of the new key. This is helpful if we intended
```

Comment:
Hm, the word "representation" is confusing to me. I don't think I'd understand this comment without the context in this PR. This makes more sense to me:
```suggestion
   * This method can construct a new key given the old key and the
   * desired contents of the new key. This is helpful if we intended
```
but I guess "contents" means different things to different people. The number of calls to this method are relatively few so it's probably fine if the comment isn't crystal clear.